dlt/extract/extract.py

```diff
@@ -193,9 +205,38 @@ def _write_dynamic_table(self, resource: DltResource, item: TDataItem) -> None:
     def _write_static_table(self, resource: DltResource, table_name: str, items: TDataItems) -> None:
         existing_table = self.dynamic_tables.get(table_name)
         if existing_table is None:
-            static_table = resource.compute_table_schema()
-            static_table["name"] = table_name
-            self.dynamic_tables[table_name] = [static_table]
+            if not self._add_dynamic_table(resource, table_name=table_name):
+                return
+        self._write_item(table_name, resource.name, items)
+
+    def _add_dynamic_table(self, resource: DltResource, data_item: TDataItem = None, table_name: Optional[str] = None) -> bool:
```

Comment:
This will have to live here somewhere, as we need to be able to intercept the creation of new items if the table is not allowed. If you don't want the extraction code mixed up with the contracts stuff, we could subclass the base extractor and put this code there somewhere or alternatively inject some kind of function in extract_with_schema that deals with this. I am not quite sure myself how to do this more elegantly. Of course we could also have the full extraction run the way it was before and then clean up at the end, but then we would potentially write files that will be discarded afterwards.